homework_solution.md

```diff
@@ -0,0 +1,474 @@
+# Part 1: Missing Station Data
+
+1. _**Write a query that fills out this table. Try your best to pick the correct station name for each ID. You may have to make some manual choices or editing based on the inconsistencies we've found. Do try to pick the correct name for each station ID based on how popular it is in the trip data.**_
+
+   ### Steps
+
+   1. Create an index on `from_station_id` and `to_station_id` on `trips` table to help query on those
+
+      ```
+      CREATE INDEX idx_from_station ON trips USING btree (from_station_id);
+      ```
+
+      ```
+      CREATE INDEX idx_to_station ON trips USING btree (to_station_id);
+      ```
+
+   2. Delete the exist `stations` table and recreate it with serial ids, so we can rely on auto-increment. (We could try and alter the table, but that seems to be an involved process, and our table being currently empty, it's easier to just recreate it)
+
+      ```
+      DROP TABLE stations;
+      ```
+
+      ```
+      CREATE TABLE stations
+      (
+         id SERIAL,
+         name VARCHAR(100)
+      );
+      ```
+
+      The effect of this can be perceived by running `SELECT nextval(pg_get_serial_sequence('stations', 'id'));`.
+      On the old table, that sequence, though present, wouldn't return a value. Now it does.
+
+   3. Get the first name for stations, assuming that as the correct one where there's more than one, combining the `from` and `to` column results
+
+      ```
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id
+      UNION
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id;
+      ```
```

Comment:
[comment]
I love how you've formatted this query. Very easy to read :+1:

Reply:
[question]
What would the output of this query look like?

Reply:
It brings the station id, and the first record in the aggregation of station names by that id. So, something like:
```
 from_station_id |             array_agg             
-----------------+-----------------------------------
            7094 | Chester Station
            7327 | Davenport Rd / Oakwood Rd - SMART
            7109 | Seaton St / Dundas St E
            7207 | Dundas St W / Crawford St
            7272 | Yonge St / Dundonald St - SMART
            7369 | Dupont St / Emerson Ave SMART
            7157 | Dufferin St / Bloor St
            7090 | Danforth Ave / Lamb Ave
            7381 | Roehampton Ave / Yonge St
            7357 | Lakeshore Blvd E /Leslie St
```

Also, for the record, this is what it'd look like if we didn't get just the first record:
```
 from_station_id |                            array_agg                             
-----------------+------------------------------------------------------------------
            7000 | {"Fort York  Blvd / Capreol Crt","Fort York  Blvd / Capreol Ct"}
            7001 | {"Lower Jarvis St / The Esplanade"}
            7002 | {"St George St / Bloor St W","St. George St / Bloor St W"}
            7003 | {"Madison Ave / Bloor St W"}
            7004 | {"University Ave / Elm St"}
            7005 | {"King St W / York St","University Ave / King St W"}
            7006 | {"Bay St / College St (East Side)"}
            7007 | {"College St / Huron St","College St W / Huron St"}
            7008 | {"Wellesley St / Queen's Park Cres"}
            7009 | {"King St E / Jarvis St"}
```